### PR TITLE
Add tests for batch indexing annotations.

### PR DESCRIPTION
**Pull request by sheetaluk**
Also a minor refactor to make the following variables injectable:
ES_CHUNK_SIZE: the number of docs in one chunk sent to ES
PG_WINDOW_SIZE: how often to log how many annotations have been logged. Also used in the query for loading all annotations in batches

### CODE REVIEW

File: h/search/index.py [hunk 1]
```diff
@@ -107,7 +107,7 @@ def __init__(self, session, es_client, request, target_index=None, op_type='inde
         else:
             self._target_index = target_index
 
-    def index(self, annotation_ids=None):
+    def index(self, annotation_ids=None, windowsize=PG_WINDOW_SIZE, chunk_size=ES_CHUNK_SIZE):
```
Comment (robertknight) [inline h/search/index.py]: Could you update the docstring for this method to explain what the `windowsize` and `chunk_size` parameters are for.

File: h/search/index.py [hunk 2]
```diff
@@ -118,17 +118,16 @@ def index(self, annotation_ids=None):
...
-            annotations = _all_annotations(session=self.session,
-                                           windowsize=PG_WINDOW_SIZE)
+            annotations = _all_annotations(session=self.session, windowsize=windowsize)
         else:
             annotations = _filtered_annotations(session=self.session,
                                                 ids=annotation_ids)
 
         # Report indexing status as we go
-        annotations = _log_status(annotations, log_every=PG_WINDOW_SIZE)
+        annotations = _log_status(annotations, windowsize)
```
Comment (robertknight) [inline h/search/index.py]: I found it slightly clearer to read when the `log_every` argument was explicitly named rather than just relying on the position.

File: tests/h/search/index_test.py [hunk 3]
```diff
@@ -310,6 +311,66 @@ def test_annotation_is_marked_deleted(self, es_client, factories, index, search)
         assert result.get('_source').get('deleted') is True
 
 
+class TestBatchIndexer(object):
+    def test_it_indexes_all_annotations(self, db_session, es_client, factories, pyramid_request):
+        annotations = factories.Annotation.create_batch(3)
+        ids = [a.id for a in annotations]
+
+        h.search.index.BatchIndexer(db_session, es_client, pyramid_request, es_client.index).index()
```
Comment (robertknight) [inline tests/h/search/index_test.py]: The `BatchIndexer` construction is repeated with the same arguments in several test cases. I would suggest moving into a fixture (say `batch_indexer` that returns an instance of `BatchIndexer`).

File: tests/h/search/index_test.py [hunk 4]
```diff
@@ -310,6 +311,66 @@ def test_annotation_is_marked_deleted(self, es_client, factories, index, search)
...
+                result = es_client.conn.get(index=es_client.index,
+                                            doc_type="annotation",
+                                            id=id)
+
+    def test_it_returns_errored_annotation_ids(self, db_session, es_client, factories, pyramid_request):
+        annotations = factories.Annotation.create_batch(3)
+        expected_errored_ids = set([annotations[0].id, annotations[2].id])
+
+        elasticsearch1.helpers.streaming_bulk = mock.Mock()
+        elasticsearch1.helpers.streaming_bulk.return_value = [(False, {'index': {'error' : 'some error', '_id': annotations[0].id}}),
```
Comment (robertknight) [inline tests/h/search/index_test.py]: Some of these lines are pretty long, which makes them annoying to read in GitHub's diff viewer:

screenshot 2018-05-31 12 04 17

Please keep them to max ~100 chars.

Comment (sheetaluk) [reply tests/h/search/index_test.py]: Hi! yes, I have a bit of linting cleanup to do.

**Comment by sheetaluk** [comment]
Ready for your eyes, @robertknight ! I'll send a follow up PR your way for:

1. Deleted annotations should not be indexed
2. Annotations are converted to the expected JSON format for indexing
3. The AnnotationTransformEvent event is emitted so that other code can "decorate" annotations with additional info. eg. A flag is set if the annotation was posted by a NIPSA-d user.

File: tests/h/search/index_test.py [hunk 5]
```diff
@@ -149,10 +153,10 @@ def test_you_can_make_aggregations_on_user_raw(self, factories, index, search):
 
         response = search.execute()
 
-        user_bucket_1 = next(bucket for bucket in response.aggregations.user_raw_terms.buckets
-                             if bucket["key"] == "acct:someone@example.com")
-        user_bucket_2 = next(bucket for bucket in response.aggregations.user_raw_terms.buckets
-                             if bucket["key"] == "acct:Someone@example.com")
+        user_bucket_1 = next(bucket for bucket in response.aggregations.user_raw_terms.buckets if
```
Comment (robertknight) [inline tests/h/search/index_test.py]: These formatting changes appear to be unrelated to the other changes in this PR. Can you remove them from this PR?

File: tests/h/search/index_test.py [hunk 6]
```diff
@@ -310,6 +316,80 @@ def test_annotation_is_marked_deleted(self, es_client, factories, index, search)
...
+            assert result["_id"] == id
+
+        for id in ids_not_to_index:
+            with pytest.raises(elasticsearch1.exceptions.NotFoundError):
+                result = es_client.conn.get(index=es_client.index,
+                                            doc_type="annotation",
+                                            id=id)
+
+    def test_it_logs_indexing_status(self, caplog, batch_indexer, factories):
+        annotations = factories.Annotation.create_batch(200)
```
Comment (robertknight) [inline tests/h/search/index_test.py]: There is overhead to indexing annotations in Elasticsearch so, to make these tests run faster, use smaller values if possible.

File: h/search/index.py [hunk 7]
```diff
@@ -107,28 +107,31 @@ def __init__(self, session, es_client, request, target_index=None, op_type='inde
...
             self._target_index = target_index
 
-    def index(self, annotation_ids=None):
+    def index(self, annotation_ids=None, windowsize=PG_WINDOW_SIZE, chunk_size=ES_CHUNK_SIZE):
         """
         Reindex annotations.
 
         :param annotation_ids: a list of ids to reindex, reindexes all when `None`.
         :type annotation_ids: collection
+        :param windowsize: how often to log indexing progress
```
Comment (robertknight) [inline h/search/index.py]: I think this docstring could be clearer about the meaning of `windowsize`. It means the number of annotations to index in between progress log statements?

File: tests/h/search/index_test.py [hunk 8]
```diff
@@ -150,9 +154,9 @@ def test_you_can_make_aggregations_on_user_raw(self, factories, index, search):
         response = search.execute()
 
         user_bucket_1 = next(bucket for bucket in response.aggregations.user_raw_terms.buckets
-                             if bucket["key"] == "acct:someone@example.com")
+            if bucket["key"] == "acct:someone@example.com")
```
Comment (robertknight) [inline tests/h/search/index_test.py]: The change in indentation here causes a lint error on my system, although not one caught by the `make lint` command. Can you change this line and the one below back to the way it was?

File: tests/h/search/index_test.py [hunk 9]
```diff
@@ -179,8 +183,10 @@ def test_you_can_make_aggregations_on_tags_raw(self, es_client, factories, index
 
         response = search.execute()
 
-        tag_bucket_1 = next(bucket for bucket in response.aggregations.tags_raw_terms.buckets if bucket["key"] == "Hello")
-        tag_bucket_2 = next(bucket for bucket in response.aggregations.tags_raw_terms.buckets if bucket["key"] == "hello")
+        tag_bucket_1 = next(bucket for bucket in response.aggregations.tags_raw_terms.buckets if
```
Comment (robertknight) [inline tests/h/search/index_test.py]: These indentation changes are not related to the rest of the PR. These lines were a bit long but I would prefer the change to be in a separate commit or PR.

File: tests/h/search/index_test.py [hunk 10]
```diff
@@ -310,6 +316,75 @@ def test_annotation_is_marked_deleted(self, es_client, factories, index, search)
...
+                                   id=id)
+
+    def test_it_logs_indexing_status(self, caplog, batch_indexer, factories):
+        annotations = factories.Annotation.create_batch(10)
+        ids = [a.id for a in annotations]
+
+        with caplog.at_level(logging.INFO):
+            batch_indexer.index(ids, 3)
+
+        for record in caplog.records:
```
Comment (robertknight) [inline tests/h/search/index_test.py]: This test still passes if you comment out the `log.info` statement in `h/search/index.py` because in that case `caplog.records` is empty.

File: tests/h/search/index_test.py [hunk 11]
```diff
@@ -310,6 +312,80 @@ def test_annotation_is_marked_deleted(self, es_client, factories, index, search)
...
 
+class TestBatchIndexer(object):
+    def test_it_indexes_all_annotations(self, batch_indexer, es_client, factories):
+        annotations = factories.Annotation.create_batch(3)
+        ids = [a.id for a in annotations]
+
+        batch_indexer.index()
+
+        results = es_client.conn.search()['hits']['hits']
+        assert [ann.get("_id") for ann in results].sort() == ids.sort()
```
Comment (robertknight) [inline tests/h/search/index_test.py]: This assert doesn't do what you expected because the `sort()` API mutates the array and then always returns `None`. Same gotcha that Hannah [encountered here](https://github.com/hypothesis/h/pull/5081#discussion_r196680399). You can use `sorted(list)` to get a sorted copy of a list.

File: tests/h/search/index_test.py [hunk 12]
```diff
@@ -310,6 +312,80 @@ def test_annotation_is_marked_deleted(self, es_client, factories, index, search)
...
+    def test_it_indexes_specific_annotations(self, batch_indexer, es_client, factories):
+        annotations = factories.Annotation.create_batch(5)
+        ids = [a.id for a in annotations]
+        ids_to_index = ids[:3]
+        ids_not_to_index = ids[3:]
+
+        batch_indexer.index(ids_to_index)
+
+        results = es_client.conn.search()['hits']['hits']
+        assert [ann.get("_id") for ann in results].sort() == ids_to_index.sort()
```
Comment (robertknight) [inline tests/h/search/index_test.py]: Same issue with `sort()` here as above.

**Comment by hmstepanek** [comment]
Oopsy...good catch! I’ll fix this this afternoon. 🙈

File: tests/h/search/index_test.py (hunk 11 above)
Comment (hmstepanek) [reply tests/h/search/index_test.py]: Sorry @sheetaluk! My bad!
